MatrixSDK/Data/MXRoom.m

```diff
@@ -824,14 +848,17 @@ - (MXHTTPOperation*)setIsDirect:(BOOL)isDirect
             directUserId = mxSession.myUser.userId;
         }
         
-        NSMutableDictionary *directRooms = [NSMutableDictionary dictionaryWithDictionary:mxSession.directRooms];
-        NSMutableArray *roomLists = (directRooms[directUserId] ? [NSMutableArray arrayWithArray:directRooms[directUserId]] : [NSMutableArray array]);
-        
-        [roomLists addObject:self.roomId];
-        
-        [directRooms setObject:roomLists forKey:directUserId];
+        @synchronized (mxSession.directRooms)
+        {
+            NSMutableArray *roomLists = (mxSession.directRooms[directUserId] ? [NSMutableArray arrayWithArray:mxSession.directRooms[directUserId]] : [NSMutableArray array]);
+            
+            [roomLists addObject:self.roomId];
+            
+            [mxSession.directRooms setObject:roomLists forKey:directUserId];
+        }
         
-        return [mxSession setDirectRooms:directRooms success:success failure:failure];
+        // Let the mxSession post the 'kMXSessionDirectRoomsDidChangeNotification' notification on account data update
```

Comment:
Other MXRoom methods does such check/wait for the caller, ie the caller does not need to call a method and then listen to a notification to get the result.